sympy/core/exprtools.py

```diff
@@ -1414,156 +1288,614 @@ def factor_nc(expr):
         return expr.func(*[factor_nc(a) for a in expr.args])
     expr = expr.func(*[expand_power_exp(i) for i in expr.args])
 
-    from sympy.polys.polytools import gcd, factor
-    expr, rep, nc_symbols = _mask_nc(expr)
+    return _NCPoly(expr).factor_one()
 
-    if rep:
-        return factor(expr).subs(rep)
-    else:
-        args = [a.args_cnc() for a in Add.make_args(expr)]
-        c = g = l = r = S.One
-        hit = False
-        # find any commutative gcd term
-        for i, a in enumerate(args):
-            if i == 0:
-                c = Mul._from_args(a[0])
-            elif a[0]:
-                c = gcd(c, Mul._from_args(a[0]))
+
+def _nc_generators(expr):
+    """Return noncommutative polynomial generators of an expression.
+
+    Commutative symbols are excluded, and functions of generators other than
+    positive integer powers are treated as generators, as with the polynomials module
+    """
```

Comment:
I wonder how we can treat negative powers here. Does the cited algorithm handle those?

Note that Poly can include negative powers e.g.:
```python
In [4]: Poly(x + 1/x)
Out[4]: Poly(x + (1/x), x, 1/x, domain='ZZ')
```
In the case of factorisation negative powers do not need to be handled because we can always transfer them to an overall numerator/denominator:
```python
In [5]: cancel(x + 1/x)
Out[5]: 
 2    
x  + 1
──────
  x 
```
The same would not work in the non-commutative case. We could factor out negative powers on the left or the right but not from the middle of a monomial like `A*B**-1*C`.

Reply:
`Poly` only *sort* of handles negative powers:
```
>>> from sympy.abc import a
>>> factor(Poly((a + 1)*(a**-1 + 1)))
Poly(a + (1/a) + 2, a, 1/a, domain='ZZ')
```
(If it did do the canceling that you mention above, then it should factor to `(a+1)**2/a` but it doesn't)
It basically treats `1/a` as if it were some separate variable `b` and doesn't know about any connection between `a` and `1/a`.

My code attempts to do the same thing: any noncommutative expression that isn't `Mul`, `Add`, or a positive integer power is added as an independent generator. Although I should update it to e.g. handle `a**-2`; right now `a**-2` would be treated as a generator but clearly the correct thing to do is to add `a**-1` as the generator if there is a negative integer power.